0000-settlement-engines/0000-settlement-engines.md

```diff
@@ -0,0 +1,404 @@
+
+## Settlement
+
+A **settlement** is the irrevocable discharge of a liability by providing something of value to the party owed.
+
+Settlements may occur on a **settlement system**, or medium for exchanging value. Some settlements may transfer funds on a **ledger**, or registry of account balances and/or transactions, which is a type of settlement system. (Although not all settlement systems are ledgers, here, the terms are sometimes used interchangeably.)
+
+Examples of settlement systems include:
+
+TODO Are these actually settlement systems?
+
+- Bank clearing houses
+- Credit card processors
+- Money transfer services
+- Cryptocurrencies, blockchains, and distributed ledgers
+- Payment channels and layer 2 networks
+- Cash or physical delivery of assets
+
+## Double-entry bookkeeping
+
+TODO Evan: Is it correct to refer to it as "credit" if you were prefunding or settling things right away? (It might be, but then it would be worth clarifying the different uses of credit -- IOUs, "credits vs debits", and this usage)
+
+TODO Explain clearing vs settlement?
+
+Together, a settlement engine and an accounting system interface with one another to perform double-entry bookkeeping. To ensure accurate, balanced double-entry bookkeeping, settlement engine and accounting system implementations MUST enforce several invariants.
+
+### Accounting system invariants
+
+#### Account for outgoing settlements
+
+The accounting system is responsible for triggering outgoing settlements. For example, when the accounts payable reaches a particular threshold, the accounting system SHOULD trigger a settlement to reduce the amount owed to the peer to a predefined, lesser amount in order to be able to continue transacting with the peer.
+
+If the accounting system opts to trigger a settlement:
+
+1. The accounting system MUST debit the accounts payable, subtracting the amount of the settlement.
+2. Then, the accounting system MUST send a request to the settlement engine to settle the amount.
+
+If the settlement engine responds that it only queued a partial amount for settlement (due to lesser precision), the accounting system MUST credit back the accounts payable, adding the leftover amount.
+
+If request retries fail per the [idempotence behavior](#idempotence), the accounting system MUST credit back the accounts payable, adding the amount of the failed settlement.
+
+#### Account for incoming settlements
+
+If the settlement engine instructs the accounting system a settlement was received, the accounting system MUST credit the accounts receivable, subtracting the amount of the settlement.
+
+The accounting system MUST respond with the amount it credited to the account. If it only credited a partial amount (due to lesser precision), the settlement engine tracks the leftover, uncredited amount and includes it in the next incoming settlement notification to the accounting system.
```

Comment:
This must be updated to say that the accounting system is responsible for tracking leftovers for incoming settlements. In that case, I think that the accounting system no longer needs to respond with the amount it credited.

Reply:
I removed this, but didn't add mention any "leftovers" in the accounting system -- I felt it was clear enough. If you feel otherwise, we could add a mention that the accounting system needs to track balances in arbitrary precision